### ISSUE TITLE
Library and Program in the same repository...

### ISSUE THREAD

**Issue by sean-** [kardianos/govendor#141]
In hashicorp/serf#386 someone identified a problem re: the way a repository can contain a program and a library.  I've seen your prior posts on go-nuts@, and would be interested in seeing if there's a way that `govendor` could possibly handle this scenario.  Below is a test that we have to verify if something can compile.  If there was a way to export a series of "named vendor sets" that `govendor` can could use to conditionally import different `vendor/` files so that a single repository could be used for either a library or a program, I think that'd work out well for everyone.  I haven't thought through the implementation details, but is this use case something `govendor` would possibly tackle?

```go
package binlib_canary_test

import (
	"log"
	"testing"

	"github.com/hashicorp/memberlist"
	"github.com/hashicorp/serf/serf"
)

type filter struct{}

func (t filter) NotifyAlive(peer *memberlist.Node) error {
	log.Println(peer.Name, peer.Addr.String(), int(peer.Port))
	return nil
}

func NewClusterOptions(nodeID, addr string, port int) Options {
	events := make(chan serf.Event)
	config := serf.DefaultConfig()
	config.MemberlistConfig = memberlist.DefaultWANConfig()
	config.MemberlistConfig.Alive = filter{}
	config.MemberlistConfig.BindAddr = addr
	config.MemberlistConfig.BindPort = port

	if nodeID != "" {
		config.NodeName = nodeID
	}

	return Options{
		Config:  config,
		EventCh: events,
	}
}

// Options holds the options for the cluster.
type Options struct {
	*serf.Config
	EventCh chan serf.Event
}

func TestSerfMemberlist(t *testing.T) {
	t.Log("success")
}
```

**Comment by kardianos** [comment]
There has been recent discussion regarding this, most recently that comes to mind is the coreos/etcd issue.

I'm very much on the fence on this proposal, but I put it out here: https://github.com/golang/go/issues/15369 . In essence it is designed to handle the case of a CMD and  lib in the same repo essentially, though I dislike that wording in go. I'll re-read this issue again when I have more time and inspect the code to ensure I didn't miss something.

**Comment by kardianos** [comment]
@sean- This might be a good idea. 

Here is my current take away right now, please correct my understanding. In ".gitignore" you would have "vendor/*/". The "vendor.json" file

**Comment by myitcv** [comment]
@sean- I don't think vendoring by itself can solve this issue. Indeed I responded to @kardianos's proposal because there has been another discussion about this [here](https://github.com/golang/go/issues/15201) in which I set forward a slightly different approach. I'm just in the process of finalising that proposal which I will shortly post via the [Go proposal process](https://github.com/golang/proposal)

**Comment by kardianos** [comment]
@sean- If I understand correctly, you might copy vendor files into folders not named "vendor", then after you pull, you might copy files from one of those other directory into vendor.

Kind of like a local version of "sync". After git pull, you might run "govendor pop vendor-release". After make a change to the vendor folder, you might run "govendor push vendor-release". And in your ".gitignore" file you would have "vendor/*/".

Please correct my understanding of this proposal.

**Comment by sean-** [comment]
That's correct. I didn't want to prime the discussion upfront, but the issue is threefold: 1) vendor/ is a monolithic fossilization of dependencies for one variant or flavor of compilation, 2) the go build system doesn't support a method of selecting which fossils are required for a given variant and assumed that each repository is used for only one variant, and 3) dependencies *should* never conflict (e.g. Conflicting types shouldn't happen). And 4) a future solution in Go won't be ratified until 1.8+, so this problem will exist for a while and a pragmatic solution that takes the existing constraints is a prudent path forward (it would be nice if this problem was tacked by Go directly).  That having been said...

Without namespaces or pre-processors, the best thing that can happen is to teach external tooling how to create "vendor sets" that parallel build variants and would continue the fossilization of dependencies, but with one layer of indirection. Something like `vendor-{bin,lib}/` would constitute two vendor sets, `bin` and `lib`, respectively, and could be hard linked or copied to `vendor/` easily.

`vendor/` would be contain the default vendor set for backwards compatibility and to preserve an "out of the box" experience where things just work. `govendor variant list` would list the available variants and `govendor variant set bin` would switch from the default set to the bin set as needed. Libs should be the default, and any binary will require a command be run upfront (not problematic if a Makefile is used).

I couldn't come up with anything that didn't mutate `git status` or didn't involve copying files or a `(sym|hard)link`, but think for the time being we're stuck with that reality. Similarly, I was not concerned with duplicate data in `vendor*/`.

?

**Comment by kardianos** [comment]
@sean- I agree with the analysis of the problem and think this design is good and worthwhile pursuing. I have some issues to resolve first, but then of all the feature requests, this is highest on my priority.

**Comment by myitcv** [comment]
@sean- in case it's of interest, https://github.com/myitcv/go-symlink sets out an approach by which a repository containing binaries and libraries can be made to work for:

* users of the binaries
* users of the libraries
* developers of the containing repository

**Comment by kardianos** [comment]
@myitcv Try cloning that on windows.

**Comment by myitcv** [comment]
@kardianos I know that's an issue (just updated the [`README.md`](https://github.com/myitcv/go-symlink#getting-started-as-a-developer-of-this-repository) to make clear the current platform limitations), hence [my proposal](https://docs.google.com/document/d/1n5y3mZPs_4PjI80a0vZEaHLe7r9PeiiE9xsIrQFT8Is/edit)

I realise this doesn't provide a cross-platform solution as of right now...

**Comment by kardianos** [comment]
@sean- 
I'm concerned about make a (relatively) complex tool more complex, but here is my idea:
 * New sub command `govendor stash save <name>` - copies vendor folder to "_vendor-<name>".
 * New sub command `govendor stash list`  - list all directories matching "_vendor-*"
 * New sub command `govendor stash pop <name>` - overwrites vendor folder from "_vendor-<name>".
 * Other sub commands might grow a argument "-name" to specify which vendor folder to use. (less sure about his one).

Thoughts, feelings?

**Comment by sean-** [comment]
@kardianos I like the parallel you're drawing between `git-stash` and `govendor`.  I think the two most common operations will be `apply <name>` and `save <name>`.  To be clear, I think the `govendor notify` command is more important than this, but I do think it would be a net-net win for the community of `govendor` had support for the bin/lib vendor problem.

As for the subdirectories, were you thinking `_vendor-<name>` or `vendor-<name>`?  I'm biased toward the latter because it lexically sorts adjacent to `vendor/` and provides a more fluid experience and hint that something is happening (or `vendor/.stash-<name>/` if fully encapsulated is desired).

Regarding the workflow, I think until the use and workflow becomes clear, I would make it impossible to have `govendor` do any harm.  By that I mean `govendor` won't trounce on `vendor/` unless a user moves it out of the way.  Something like:

1. `govendor stash save --current --name=default` == Moves `vendor/` to `vendor-default/` *if* `vendor-default/` doesn't exist.
2. `govendor stash load --name=default` == Copies `vendor-default/` to `vendor/`
3. `govendor stash list` == Shows the list of `vendor-*/` stashes

In typing that out, I think `save`/`load` is probably more natural than `pop`/`save`.

On *NIX platforms maybe detect that `vendor` is a symlink pointing to `vendor-default/` so there's no copying involved?  In an ideal world `vendor/` is a symlink and added to `.gitignore` so libraries have nothing out of the box in `vendor/` but binaries can ship with `vendor/` pre-setup.

?

**Comment by kardianos** [comment]
Good thoughts all. I'll like your thinking on these.
I prefer `_vendor-<name>` because then the std tools, including govendor itself, won't descend into
 those directories.

**Comment by sean-** [comment]
...and that is an excellent point.  :~]  It may be worth noting that in a comment somewhere (I totally forgot about that rule but remember it from `godep`).

**Comment by sean-** [comment]
I just learned about [manul](https://github.com/kovetskiy/manul) and think it's approach using `git` submodules is a clever-enough way to swap in and out sets of vendored dependencies.  Generally speaking I loath `git` submodules but named sets of submodules managed via `govendor` might be one way to augment what is shipped in `vendor/`. I need to chew on this for a while, but thought I'd pass it along.

**Comment by kardianos** [comment]
There is talk about maybe altering vendoring in some manner to truly address this.
@nf may represent the go core team. So for now I'd like to hold off on this idea.

But regardless of versioning or metadata stuff, this is what really is going to matter.
